include/picongpu/plugins/binning/Axis.hpp

```diff
@@ -0,0 +1,317 @@
+             */
+            template<typename T_Data>
+            class Range
+            {
+            public:
+                T_Data min; ///< Minimum of binning range
+                T_Data max; ///< Maximum of binning range
+                Range(T_Data minIn, T_Data maxIn) : min{minIn}, max{maxIn}
+                {
+                    PMACC_VERIFY(min < max);
+                }
+            };
+
+            /**
+             * @brief Holds the axis range and information on how this range is split into bins
+             */
+            template<typename T_Data>
+            class AxisSplitting
+            {
+            public:
+                Range<T_Data> m_range; ///< Range object
+                uint32_t nBins; ///< Number of bins in range
+                AxisSplitting(Range<T_Data> range, uint32_t numBins) : m_range{range}, nBins{numBins}
+                {
+                }
+            };
+
+            // how to check if particle has this attribute? at compile time. throw error
+
+            // Overflow bins is the resposibilty of the axis implementation. Make nBins = user_defined_n_bins + 2
+            // Bins which don't have an inherent size should return a bin width of 1 to not influence normalization.
+
+            // TODO enforce some sort of interface on axis
+            // TODO mark functions which are mandatory for each type of axis
+
+
+            /**
+             * Linear axis with contiguous fixed sized bins.
+             * Axis splitting is defined with min, max and n_bins. Bin size = (max-min)/n_bins.
+             * Bins are closed open [) intervals [min, min + size), [min + size, min + 2*size) ,..., [max-size, max).
+             * Allocates 2 extra bins, for under and overflow. These are bin index 0 and (n_bins+2)-1
+             */
+            template<typename T_Attribute, typename T_AttrFunctor>
+            class LinearAxis
+            {
+            public:
+                using T = T_Attribute;
+                AxisSplitting<T_Attribute> axisSplit;
+                std::string label; ///> Axis name, written out to OpenPMD
+                std::array<double, 7> units; ///> Units(Dimensionality) of the axis
+                // TODO store edges? Copmute once at the beginning and store for later to print at every iteration,
+                // also to be used in search based binning
+                std::vector<double> binEdges;
+                struct LinearAxisKernel
+                {
+                    T_AttrFunctor
+                        getAttributeValue; ///< Function to place particle on axis, returns same type as min and max
+                    T min, max; /**< Min and max values in the range of the binning. Values outside this range are
+                                placed in overflow bins */
+                    uint32_t nBins; ///< Number of bins in range
+                    using ScalingType = std::
```

Comment:
please add a comment that we use a type that avoids precision loss.

Reply:
added